### PR TITLE
ROS2: added launch file and changes necessary for it to work

### PR DESCRIPTION
**Pull request by klintan**
From this PR https://github.com/ros-perception/image_pipeline/pull/443

- Added launch file using components
- Added necessary changes for it to work
- All parameters kept,
- Camera namespace cmd-line argument still available and same behavior.

@SteveMacenski and @JWhitleyAStuff I would love some feedback on this, and since you saw the last one I think it makes sense for you to review this again.

### CODE REVIEW

**Review by JWhitleyWork — APPROVED** [review]
I'm not familiar enough yet with composable nodes to definitively say much but it looks good as far as I'm concerned. I'll wait for @SteveMacenski to review before merging.

**Comment by SteveMacenski** [comment]
There’s a little too much pending on the ros2 side here for me to merge anything. I’m working on allocating time post-roscon to pull away from my other work and refocus my attention here. The ROS2 branches have gotten out of wack with ROS1 & there’s been alot of non-scrutinized changes that need to be looked at again

**Comment by SteveMacenski** [comment]
Things I know I’m not going to like when I review it
- any static variables. Namespace, amongst others should be runtime configurable via the launch configuration options
- debug main should be removed unless there’s a test using it
- are these composible nodes? I don’t recall that

File: image_proc/src/image_proc.cpp
Comment (SteveMacenski) [inline image_proc/src/image_proc.cpp:78]: Its not clear to me what your refactor buys us vs the old changes of declaring the parameters outside the node itself and here in the main()

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
More of a question, but if you've tested this it looks good to me.

There's the open question above about runtime configurations to change the namespace though. See `LaunchConfiguration`

File: image_proc/src/image_proc.cpp
Comment (klintan) [reply image_proc/src/image_proc.cpp:78]: I'm not sure how one would do it if we want to use it as a component. Before, trying to use it as a component, the callback was never triggered and the camera_namespace would not be declared.

File: image_proc/src/rectify.cpp
Comment (SteveMacenski) [inline image_proc/src/rectify.cpp:82]: Can we set these as logical defaults? image_color, image_mono, camera_namespace being camera, etc. Right now if you basically dont use the launch file it'll never work

File: image_proc/src/rectify.cpp
Comment (SteveMacenski) [inline image_proc/src/rectify.cpp:71]: I think this should just go back to the way it was

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: image_proc/src/image_proc.cpp
Comment (SteveMacenski) [reply image_proc/src/image_proc.cpp:78]: mhm. So I'm trying to rectify 2 things, both the use in the lanuch file that you have so you can do commandline remapping and the use of the nodes as individual components. Right now if you don't use this main() for the library, you really cant use the components themselves because the declares don't set logical defaults and they don't try to get the parameter themselves. 

You can set arguments through the NodeOptions class http://docs.ros2.org/dashing/api/rclcpp/classrclcpp_1_1NodeOptions.html#ab0c61163fb466ceb980a001bdce2fb7a

**Comment by JWhitleyWork** [comment]
@klintan Any updates? Also needs rebase.

**Comment by klintan** [comment]
> @klintan Any updates? Also needs rebase.

Yes sorry I moved on to other things and sort of forgot about this one, I'll try to take a look this weekend and rebase + fix feedback. Might be a better way to do it now.

**Comment by JWhitleyWork** [comment]
Ping @klintan

**Comment by klintan** [comment]
Sorry for the delay.

I'll try to get to this during the holidays :) 

Thanks for reminding me @JWhitleyWork

**Comment by pmusau17** [comment]
Ping @klintan

**Comment by clalancette** [comment]
I'm going to close this one out since there has been no response in a long time.  Please feel free to reopen if you are planning on updating this PR.

**Pull request closed without merging**
